codewof/users/views.py

```diff
@@ -198,3 +219,378 @@ class UserAPIViewSet(viewsets.ReadOnlyModelViewSet):
+class GroupDeleteView(LoginRequiredMixin, AdminRequiredMixin, DeleteView):
+    """View for deleting a group."""
+
+    model = Group
+
+    def get_success_url(self):
+        """URL to route to on successful delete."""
+        return reverse('users:dashboard')
+
+
+def admin_required(f):
+    """Decorator for checking the user is an Admin of the Group."""
+
+    @wraps(f)
+    def g(request, *args, **kwargs):
+        admin_role = GroupRole.objects.get(name="Admin")
+        group = Group.objects.get(pk=kwargs['pk'])
+        if Membership.objects.all().filter(user=request.user, group=group, role=admin_role):
+            kwargs['group'] = group
+            return f(request, *args, **kwargs)
+        else:
+            raise PermissionDenied()
+
+    return g
+
+
+@require_http_methods(["PUT"])
+@login_required()
+@admin_required
+@transaction.atomic
+def update_memberships(request, pk, group):
+    """View for updating memberships from JSON."""
+    body_unicode = request.body.decode('utf-8')
+    body = json.loads(body_unicode)
+    memberships = body['memberships']
+
+    for membership in memberships:
+        id = membership['id']
+        delete = membership['delete']
+        role = membership['role']
+
+        if type(id) != int:
+            raise Exception("One of the membership objects has an id that is not an integer (id={}).".format(id))
+        if type(delete) != bool:
+            raise Exception("One of the membership objects has delete value that is not a boolean (id={}).".format(id))
+
+        membership_object = Membership.objects.filter(id=id)
+        if len(membership_object) == 0:
+            raise ObjectDoesNotExist
+        else:
+            membership_object = membership_object[0]
+
+        if delete:
+            membership_object.delete()
+        else:
+            role_object = GroupRole.objects.filter(name=role)
+            if len(role_object) == 0:
+                raise Exception("One of the membership objects has a non-existent role (id={}).".format(id))
+            else:
+                membership_object.role = role_object[0]
+                membership_object.save()
+
+    if len(Membership.objects.filter(group=group, role=GroupRole.objects.get(name='Admin'))) == 0:
+        raise Exception("Must have at least one Admin in the group.")
+
+    return HttpResponse()
+
+
+class SufficientAdminsMixin:
+    """Mixin for checking there will be enough Admins if deleting an Admin Membership."""
+
+    def dispatch(self, request, *args, **kwargs):
+        membership = self.get_object()
+        admin_role = GroupRole.objects.get(name='Admin')
+        if len(Membership.objects.all().filter(group=membership.group,
+                                               role=admin_role)) > 1 or membership.role != admin_role:
+            return super().dispatch(request, *args, **kwargs)
+        else:
+            raise Exception("A Group must have at least one Admin.")
+
+
+class RequestUserIsMembershipUserMixin:
```

Comment:
If you are finding the number of Mixins hard to deal with, you are welcome to put them in a separate `mixins.py` file in the same directory.

Reply:
Good idea, done